### PR TITLE
[core] Test some XPath corner cases

### LINKED ISSUE #1939: [core] XPath expressions return handling
XPathRule doesn't check the type of the nodes returned by an XPath expression:

https://github.com/pmd/pmd/blob/2d4a30fa106b1fa35eb1849b7a97a4a1ee1a35f9/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/SaxonXPathRuleQuery.java#L86-L95

The assignment of the result of `xpathExpression::evaluate` to `List<ElementNode>` line 86 is unchecked, it's actually a list of the more general `Item`. So when getting a node line 93, the routine will fail with `ClassCastException` if the element is not actually an `ElementNode`.

This is not completely outlandish: eg `1+1` is a valid XPath expression that doesn't yield an ElementNode, and so is `/` (a single slash), which should yield a document node (see #1938) 

We should at least ensure that the type of each element is correct

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

These are fixed only for Saxon. The merge into 7.0.x will be non-trivial though.

## Related issues

- Fixes #1939
- Adds tests for #1938, to check that Saxon has the correct behavior. But this issue is a wont-fix for Jaxen. 
- Refs #2523 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/ac9add6c927147bb8e1a4b44877c4562/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: docs/pages/release_notes.md [hunk 1]
```diff
@@ -16,6 +16,10 @@ This is a {{ site.pmd.release_type }} release.
 
 ### Fixed Issues
 
+- pmd-core
```
Comment (adangel) [inline docs/pages/release_notes.md:19]:
```suggestion
*   pmd-core
```

File: docs/pages/release_notes.md [hunk 2]
```diff
@@ -16,6 +16,10 @@ This is a {{ site.pmd.release_type }} release.
...
+   - [#1939](https://github.com/pmd/pmd/issues/1939) \[core] XPath expressions return handling
```
Comment (adangel) [inline docs/pages/release_notes.md:20]:
```suggestion
    * [#1939](https://github.com/pmd/pmd/issues/1939): \[core] XPath expressions return handling
```

**Review by adangel — APPROVED** [review]
